File: C#/Calculations/SelfCheckout.cs
using static CSharp.DataValidation;

namespace CSharp.Calculations
{
    internal class SelfCheckout
    {
        private const float TAX_RATE = 0.055f;

        private class Item
        {
            private readonly float? price, quantity;
            public float? subtotal;
            public Item(int _i)
            {
                price = ValidatedNumber<float>($"Enter the price of item {_i}: ", _sentinel: "STOP");
                if (price == null)
                    throw new Exception();
                
                quantity = ValidatedNumber<float>($"Enter the quantity of item {_i}: ", _sentinel: "STOP");
                if (quantity == null)
                    throw new Exception();
                subtotal = price * quantity;
            }
        }

        private class Basket
        {
            private readonly float? subtotal, tax, total;
            public Basket(List<Item> _items)
            {
                foreach (Item item in _items)
                    subtotal += item.subtotal;
                tax = subtotal * TAX_RATE; 
                total = subtotal + tax;
            }

            public void DisplayTotals()
            {
                Console.WriteLine($"Subtotal: ${subtotal:F2}");
                Console.WriteLine($"Tax: ${tax:F2}");
                Console.WriteLine($"Total: ${total:F2}");
            }
        }

        public static void Main8()
        {
            List<Item> items = [];
            for (int i = 1; true; i++)
            {
                try { items.Add(new Item(i)); }
                catch { break; }
            }
            Basket checkout = new(items);
            checkout.DisplayTotals();
        }
    }
}

File: Python/Calculations/SelfCheckout.py
# Revise the program to ensure that prices and quantities are entered as numeric values.
# Don't allow the user to proceed if the value entered is not numeric.
from DataValidation import ValidatedNumber

# Create a simple self-checkout system.
# Prompt for the prices and quantities of three items.
# Calculate the subtotal of the items.
# Then calculate the tax using a tax rate of 5.5%.
# Print out the line items with the quantity and total,
# and then print out the subtotal, tax amount, and total.
TAX_RATE = 0.055


# Keep the input, processing, and output parts of your program separate.
# Collect the input, then do the math operations and string building,
# and then print out the output.
class Item:
    def __init__(self, _i):
        self.price = ValidatedNumber(f"Enter the price of item {_i}: ", _sentinel="STOP")
        if self.price is None:
            raise ValueError
        self.quantity = ValidatedNumber(f"Enter the quantity of item {_i}: ", _sentinel="STOP")
        if self.quantity is None:
            raise ValueError
        self.subtotal = self.price * self.quantity


class Basket:
    def __init__(self, _items):
        self.items = _items
        self.subtotal = 0.0
        for each in self.items:
            self.subtotal += each.subtotal
        self.tax = self.subtotal * TAX_RATE
        self.total = self.subtotal + self.tax

    def display_totals(self):
        print(f"Subtotal: ${self.subtotal:.2f}")
        print(f"Tax: ${self.tax:.2f}")
        print(f"Total: ${self.total:.2f}")


items = []
i = 1
while True:
    try:
        items.append(Item(i))
    except ValueError:
        break
    i += 1

checkout = Basket(items)
checkout.display_totals()
